Readme.md

```diff
@@ -0,0 +1,1432 @@
+
+This gets even more prominent when there is a need to access individual fields, for example, when iterating. Do not ever write this:
+
+```scala
+val joinedPayloadByInterface = actions.iterator
+  .map(kv => kv._1._1 -> kv._2._3).toVector
+  .groupBy(_._1)
+  .map(kv => kv._1 -> kv._2.mkString)
+```
+
+This is much better:
+
+```scala
+val joinedPayloadByInterface = actions.iterator
+  .map { case (propId, action) => propId.interface -> action.payload }
+  .toVector
+  .groupBy { case (interface, _) => interface }
+  .map { case (interface, payloads) => interface -> payloads.mkString }
+```
+
+And even this is rather complex; split the chained operations into several variable bindings:
+
+```scala
+val singlePayloadByInterface = actions.iterator
+  .map { case (propId, action) => propId.interface -> action.payload }
+  .toVector
+
+val allPayloadsByInterface = singlePayloadByInterface
+  .groupBy { case (interface, _) => interface }
+
+val joinedPayloadByInterface = allPayloadsByInterface
+  .map { case (interface, payloads) => interface -> payloads.mkString }
+```
+
+This way the intention of the code is very clear even to a relatively casual reader.
+
+# Concurrency (TODO)
+
+# Patterns and architecture
+
+Scala is a multi-paradigm language, combining elements from functional and object-oriented programming. However, this is dangerous combination which may result in a very obscure code, if from both paradigms are used indiscriminately.
+
+As a general rule, object-oriented patterns should be used to model the high-level architecture of the program, while functional patterns should be used locally, to write the actual implementation code. Think of object-oriented design as a strategy, and of functional patterns as tactics: apply the commonly known object-oriented patterns when it is necessary to structure the overall program structure, and use functional idioms when you're writing concrete code which manipulates concrete data.
+
+Of course, it is sometimes difficult to make a clean separation, for example, relying on immutability leads to creation of many ADTs, which definitely affect the high-level structure of the program, and vice versa, using dependency injection patterns have very noticeable effect on the concrete implementation. But this approach allows ruling out harmful practices like using high-level functional idioms to structure the overall program (e.g. trying to design the entire program workflow as a monad transformer stack) or blindly using object-oriented patterns to structure low level code (e.g. using virtual dispatch with a trait hierarchy where a simple match would do nicely).
+
+
+## Cake pattern
+
+See the full rule description [here][cake-pattern].
+
+In short, while the cake pattern is nice in theory, it is hard to get it right and take all the necessary things like dependencies lifetime into account. It also makes testing more difficult as the number of dependencies of different cake components grow.
```

Comment:
Yes, it may be true. However, cake pattern still remains a very Scala-specific thing (e.g. it uses self type annotations, I don't remember any other language with a similar feature), and thus it will likely make it harder for new people (especially from Java world) to be introduced into the project.

There are other problems with the cake pattern that are missing from the page I've linked to, for example, it is quite bolierplatey to manage "subcomponents" with different lifetimes (see e.g. [here](http://stackoverflow.com/questions/17525950/scala-cake-pattern-for-objects-with-different-lifetimes)).